### PR TITLE
Implemented QR code generation for authentication

### PR DESCRIPTION
**Pull request by IgorGnusFreja**

### CODE REVIEW

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 1]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
         return (System.currentTimeMillis() + pollingTimeoutInMilliseconds) < pollingEndTime;
     }
 
+    public static byte[] readAllBytes(InputStream inputStream) throws IOException {
```
Comment (milicajelic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:122]: Maybe this one can be private.

**Review by milicajelic-frejaeid — CHANGES_REQUESTED** [review]

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java [hunk 2]
```diff
@@ -168,6 +174,63 @@ public final <Response extends FrejaHttpResponse> Response send(String methodUrl
...
+                        String.format("Received unsupported HTTP status code %s. Received HTTP message: %s.",
+                                      httpResponse.getStatusLine().getStatusCode(), responseString));
+            }
+            switch (httpStatusCode) {
+                case OK:
+                case NO_CONTENT:
+                    return httpResponse;
+                case BAD_REQUEST:
+                case UNPROCESSABLE_ENTITY:
+                    FrejaHttpErrorResponse errorResponse =
```
Comment (milicajelic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:212]: In a case of Freja requests we can receive an error response from RP service that we can deserialize, but here an error can happen only if something went wrong like wrong request, resource service not working etc... So, here we can just return an exception in case of any error.

**Review by milicajelic-frejaeid — CHANGES_REQUESTED** [review]

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 3]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
...
+                while ((readLen = inputStream.read(buf, 0, bufLen)) != -1)
+                    outputStream.write(buf, 0, readLen);
+
+                return outputStream.toByteArray();
+            }
+        } catch (IOException e) {
+            exception = e;
+            throw e;
+        } finally {
+            if (exception == null) inputStream.close();
```
Comment (milicajelic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:139]: Note sure if this if is needed, also this exception logic confuses me a bit.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaResourceEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package com.verisec.frejaeid.client.enums;
+
+/**
+ * FrejaResourceEnvironment that can be used are:
+ * <br> - <b>PRODUCTION</b> - for sending HTTP requests to production
+ * environment
+ * <br> - <b>TEST</b> - for sending HTTP requests to test environment
+ */
+public enum FrejaResourceEnvironment {
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaResourceEnvironment.java:9]: Does it make sense that instead of introducing new enum, we re-use existing enum FrejaEnvironment?
Currently, FrejaEnvironment only contains address of RP service, we could expand it and add property resourceServiceUrl. That way all environment specific things are in one place.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 3 above)
Comment (mirkostojmenovic-frejaeid) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:139]: I agree, handling exceptions is a bit confusing, we can go through the code together.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 5]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
...
+            else try {
+                inputStream.close();
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:141]: This looks like a side-effect of this method. I wouldn't expect that the method closes input stream that was passed as argument, since the name of the method is readAllBytes, and also it wasn't initially opened by this method.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 6]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
...
     }
 
+    public static byte[] readAllBytes(InputStream inputStream) throws IOException {
+        final int bufLen = 4 * 0x400; // 4KB
+        byte[] buf = new byte[bufLen];
+        int readLen;
+        IOException exception = null;
+
+        try {
+            try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:129]: Seems unnecessary to have 2 try catch blocks here.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 7]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
         return (System.currentTimeMillis() + pollingTimeoutInMilliseconds) < pollingEndTime;
     }
 
+    public static byte[] readAllBytes(InputStream inputStream) throws IOException {
+        final int bufLen = 4 * 0x400; // 4KB
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:123]: Hex numbers do look interesting, but I woud argue that simple 1024 would be much more readable by human eye.
Also, you can name variable something like "bufferLengthInKB", which would make it even clearer what are the units.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java [hunk 8]
```diff
@@ -101,4 +119,30 @@ private boolean isPollingTimeExpired(long pollingEndTime) {
...
+        byte[] buf = new byte[bufLen];
+        int readLen;
+        IOException exception = null;
+
+        try {
+            try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+                while ((readLen = inputStream.read(buf, 0, bufLen)) != -1)
+                    outputStream.write(buf, 0, readLen);
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:131]: I usually use brackets even when there's only one statment in a while block. It makes it more error prone when somebody else changes code, it cannot accidentally change order of line and make something else execute in the while loop.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java [hunk 9]
```diff
@@ -72,6 +76,17 @@ public String initiate(InitiateAuthenticationRequest initiateAuthenticationReque
         return reference;
     }
 
+    @Override
+    public byte[] initiateQRCodeAuthentication(String reference) throws FrejaEidClientInternalException, FrejaEidException, IOException {
+        LOG.debug("Initiating generation of QR code for authentication from transaction reference {}.", reference);
+        String encodedReference = URLEncoder.encode(reference, StandardCharsets.UTF_8.toString());
+        String bindTransactionToUserUrl = "frejaeid://bindUserToTransaction?transactionReference=" + encodedReference;
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java:83]: I would extract prefix "frejaeid://..." to a constant.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java [hunk 10]
```diff
@@ -168,6 +174,63 @@ public final <Response extends FrejaHttpResponse> Response send(String methodUrl
...
+            httpStatusCode = HttpStatusCode.getHttpStatusCode(httpStatusCodeValue);
+            if (httpStatusCode == null) {
+                throw new FrejaEidException(
+                        String.format("Received unsupported HTTP status code %s. Received HTTP message: %s.",
+                                      httpResponse.getStatusLine().getStatusCode(), responseString));
+            }
+            switch (httpStatusCode) {
+                case OK:
+                case NO_CONTENT:
+                    return httpResponse;
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:209]: I strongly suggest having only one exit point from the method, otherwise it's very hard to follow all potential flows of execution.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/util/MethodUrl.java [hunk 11]
```diff
@@ -24,7 +24,8 @@ public enum MethodUrl {
     ORGANISATION_ID_CANCEL_ADD("/organisation/management/orgId/1.0/cancelAdd"),
     ORGANISATION_ID_GET_ALL_USERS("/organisation/management/orgId/1.0/users/getAll"),
     CUSTOM_IDENTIFIER_SET("/user/manage/1.0/setCustomIdentifier"),
-    CUSTOM_IDENTIFIER_DELETE("/user/manage/1.0/deleteCustomIdentifier");
+    CUSTOM_IDENTIFIER_DELETE("/user/manage/1.0/deleteCustomIdentifier"),
+    QR_CODE_AUTHENTICATION_GET("/qrcode/generate");
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/util/MethodUrl.java:28]: Maybe name it it says in the url, QR_CODE_GENERATE

**Review by mirkostojmenovic-frejaeid — CHANGES_REQUESTED** [review]

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 1 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:122]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/api/AuthenticationClientApi.java [hunk 12]
```diff
@@ -85,4 +86,12 @@ public AuthenticationResult pollForResult(AuthenticationResultRequest authentica
      */
     public void cancel(CancelAuthenticationRequest cancelAuthenticationRequest)
             throws FrejaEidClientInternalException, FrejaEidException;
+
+    /**
+     * Return QR code for authentication in a byte array.
+     *
+     * @param reference contains transaction reference.
+     */
+    public byte[] initiateQRCodeAuthentication(String reference)
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/api/AuthenticationClientApi.java:95]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/util/MethodUrl.java (hunk 11 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/util/MethodUrl.java:28]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java (hunk 9 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java:83]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 8 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:131]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 7 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:123]: Done 😄

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaResourceEnvironment.java (hunk 4 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaResourceEnvironment.java:9]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java (hunk 10 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:209]: Done, no_content is now treated as an error and throws an exception as it is not the expected result.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java (hunk 2 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:212]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 6 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:129]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 5 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:141]: Agreed, that part of the code was removed and the entire function simplified a bit. Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java (hunk 3 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/AuthenticationService.java:139]: Done, removed this part of the code and simplified the function.

**Review by milicajelic-frejaeid — DISMISSED** [review]

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java [hunk 13]
```diff
@@ -203,6 +208,11 @@ public GenericBuilder setTestModeCustomUrl(String serverCustomUrl) {
             return this;
         }
 
+        public GenericBuilder setTestModeResourceUrl(String resourceCustomUrl) {
```
Comment (mirkostojmenovic-frejaeid) [inline FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java:211]: **[MINOR]**
When there was only one URL that can be defined, then 'setTestModeCustomUrl' without specifying what url it is works. Would it make sense to change the method name to 'setTestModeServerCustomUrl', and also for resource service url method?

**Review by mirkostojmenovic-frejaeid — CHANGES_REQUESTED** [review]

**Review by milicajelic-frejaeid — DISMISSED** [review]

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java [hunk 14]
```diff
@@ -168,6 +174,59 @@ public final <Response extends FrejaHttpResponse> Response send(String methodUrl
...
+                                                        httpResponse.getStatusLine().getStatusCode()));
+                default:
+                    throw new FrejaEidException(String.format("HTTP code %s message: %s",
+                                                              httpResponse.getStatusLine().getStatusCode(),
+                                                              response));
+            }
+        } catch (IOException | URISyntaxException e) {
+            throw new FrejaEidClientInternalException("Failed to send HTTP request.", e);
+        } finally {
+            if (httpResponse != null && httpStatusCode != HttpStatusCode.NO_CONTENT) {
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:220]: I believe that means that the message has no body so no bytes to read from the response.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java [hunk 15]
```diff
@@ -119,6 +121,15 @@ public void cancel(CancelAuthenticationRequest cancelAuthenticationRequest)
                   cancelAuthenticationRequest.getAuthRef());
     }
 
+    @Override
+    public byte[] generateQRCodeForAuthentication(String reference) throws FrejaEidClientInternalException,
+            FrejaEidException, IOException {
+        LOG.debug("Initiating generation of byte code from transaction reference {}.", reference);
+        byte[] responseContent = authenticationService.getQRCodeContentResponse(reference);
+        LOG.debug("Received transaction response byte code.");
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java:129]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/BasicService.java [hunk 16]
```diff
@@ -24,4 +37,13 @@ protected boolean isFinalStatus(TransactionStatus status) {
                 || status == TransactionStatus.REJECTED);
     }
 
+    public byte[] getQRCodeContentResponse(String reference) throws FrejaEidClientInternalException,
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/service/BasicService.java:40]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java [hunk 17]
```diff
@@ -119,6 +121,15 @@ public void cancel(CancelAuthenticationRequest cancelAuthenticationRequest)
                   cancelAuthenticationRequest.getAuthRef());
     }
 
+    @Override
+    public byte[] generateQRCodeForAuthentication(String reference) throws FrejaEidClientInternalException,
+            FrejaEidException, IOException {
+        LOG.debug("Initiating generation of byte code from transaction reference {}.", reference);
+        byte[] responseContent = authenticationService.getQRCodeContentResponse(reference);
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/AuthenticationClient.java:128]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java [hunk 18]
```diff
@@ -186,4 +245,18 @@ final String getLibVersion() {
...
+        final int bufferLengthInKB = 1024;
+        byte[] buffer = new byte[bufferLengthInKB];
+        int line;
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+            while ((line = inputStream.read(buffer)) != -1) {
+                outputStream.write(buffer, 0, line);
+            }
+            return outputStream.toByteArray();
+        } catch (IOException ex) {
+            throw new FrejaEidClientInternalException("Failed to read bytes returned in http response", ex);
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/http/HttpService.java:259]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java [hunk 19]
```diff
@@ -109,9 +112,11 @@ public GenericBuilder(String keystorePath, String keystorePass, String certifica
         private void setServerCustomUrl(FrejaEnvironment frejaEnvironment) {
             LOG.debug("Setting Freja environment {}", frejaEnvironment == FrejaEnvironment.TEST ?
                     FREJA_ENVIRONMENT_TEST : FREJA_ENVIRONMENT_PROD);
-            serverCustomUrl = FrejaEnvironment.PRODUCTION.getUrl();
+            serverCustomUrl = FrejaEnvironment.PRODUCTION.getServiceUrl();
+            resourceServiceUrl = FrejaEnvironment.PRODUCTION.getResourceUrl();
             if (frejaEnvironment == FrejaEnvironment.TEST) {
-                serverCustomUrl = FrejaEnvironment.TEST.getUrl();
+                serverCustomUrl = FrejaEnvironment.TEST.getServiceUrl();
+                resourceServiceUrl = FrejaEnvironment.TEST.getResourceUrl();
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java:119]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaEnvironment.java [hunk 20]
```diff
@@ -11,24 +11,29 @@ public enum FrejaEnvironment {
...
+    PRODUCTION("https://services.prod.frejaeid.com", "https://resources.prod.frejaeid.com"),
     /**
      * Test address.
      */
-    TEST("https://services.test.frejaeid.com");
+    TEST("https://services.test.frejaeid.com", "https://resources.test.frejaeid.com");
 
-    private final String url;
+    private final String serviceUrl;
+    private final String resourceUrl;
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/enums/FrejaEnvironment.java:21]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java [hunk 21]
```diff
@@ -109,9 +112,11 @@ public GenericBuilder(String keystorePath, String keystorePass, String certifica
         private void setServerCustomUrl(FrejaEnvironment frejaEnvironment) {
             LOG.debug("Setting Freja environment {}", frejaEnvironment == FrejaEnvironment.TEST ?
                     FREJA_ENVIRONMENT_TEST : FREJA_ENVIRONMENT_PROD);
-            serverCustomUrl = FrejaEnvironment.PRODUCTION.getUrl();
+            serverCustomUrl = FrejaEnvironment.PRODUCTION.getServiceUrl();
+            resourceServiceUrl = FrejaEnvironment.PRODUCTION.getResourceUrl();
             if (frejaEnvironment == FrejaEnvironment.TEST) {
-                serverCustomUrl = FrejaEnvironment.TEST.getUrl();
+                serverCustomUrl = FrejaEnvironment.TEST.getServiceUrl();
```
[the remark this answers is not in the document]
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java:118]: Done.

File: FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java (hunk 13 above)
Comment (IgorGnusFreja) [reply FrejaEidClient/src/main/java/com/verisec/frejaeid/client/client/impl/BasicClient.java:211]: Done.

**Review by milicajelic-frejaeid — APPROVED** [review]

**Review by mirkostojmenovic-frejaeid — APPROVED** [review]
